### PR TITLE
cxx-qt-gen: parse invokables and methods into separate lists

### PR DESCRIPTION
**Pull request by ahayzen-kdab**

### CODE REVIEW

File: cxx-qt-gen/src/parser/qobject.rs [hunk 1]
```diff
@@ -29,12 +29,14 @@ pub struct ParsedQObject {
     pub rust_struct: Option<ItemStruct>,
     /// This is a representation of the Signals enum that defines the Q_SIGNALS for the QObject
     pub signals: Option<ParsedSignalsEnum>,
+    /// This is the list of invokables that need to be implemented on the C++ object in Rust
+    ///
+    /// These will also be exposed as Q_INVOKABLE on the C++ object
+    pub invokables: Vec<ImplItemMethod>,
     /// This is the list of methods that need to be implemented on the C++ object in Rust
```
Comment (Be-ing) [inline cxx-qt-gen/src/parser/qobject.rs:36]:
```suggestion
    /// The list of methods that need to be implemented on the C++ object in Rust
```

`This is` starts documenting every item in this struct but it adds nothing

File: cxx-qt-gen/src/parser/qobject.rs [hunk 2]
```diff
@@ -44,6 +46,36 @@ pub struct ParsedQObject {
...
+        for item in items {
+            // Check if this item is a method
+            if let ImplItem::Method(method) = item {
+                // Determine if this method is an invokable
+                if let Some(index) = attribute_find_path(&method.attrs, &["invokable"]) {
+                    // Remove the invokable attribute
+                    //
+                    // TODO: in the future there may be additional arguments
+                    // in the invokable attribute that we need to parse
+                    // such as type conversions, so we may need a ParsedInvokable later.
```
Comment (Be-ing) [inline cxx-qt-gen/src/parser/qobject.rs:63]: Does it really help to have a TODO comment for something we don't know if we will need?

File: cxx-qt-gen/src/parser/qobject.rs [hunk 3]
```diff
@@ -44,6 +46,36 @@ pub struct ParsedQObject {
     pub others: Vec<Item>,
 }
 
+impl ParsedQObject {
+    /// Extract either invokables or methods from [syn::ImplItem]'s from each Impl block
```
Comment (Be-ing) [inline cxx-qt-gen/src/parser/qobject.rs:50]:
```suggestion
    /// Extract all methods (both invokable and non-invokable) from [syn::ImplItem]'s from each Impl block
```

File: cxx-qt-gen/src/extract.rs [hunk 4]
```diff
@@ -746,13 +678,14 @@ pub fn extract_qobject(module: &ItemMod) -> Result<QObject, TokenStream> {
...
-    // Extract the normal and invokables from the Parser methods
-    let extracted = extract_invokables(&qobject.methods, &qt_ident)?;
-
     // A list of the invokables for the struct
-    let object_invokables = extracted.invokables;
+    let object_invokables = qobject
+        .invokables
+        .iter()
+        .map(|invokable| extract_invokable(invokable, &qt_ident))
+        .collect::<Result<Vec<Invokable>, TokenStream>>()?;
```
Comment (Be-ing) [inline cxx-qt-gen/src/extract.rs:686]: https://turbo.fish/::%3CResult%3CVec%3CInvokable%3E,%20TokenStream%3E%3E

Comment (Be-ing) [reply cxx-qt-gen/src/extract.rs:686]: I suspect this line would look pretty ridiculous to someone who doesn't know Rust well :sweat_smile:

Comment (ahayzen-kdab) [reply cxx-qt-gen/src/extract.rs:686]: Hehe, seems no worse than templating in C++ or Java :shrug: :-)

File: cxx-qt-gen/src/parser/qobject.rs (hunk 2 above)
Comment (ahayzen-kdab) [reply cxx-qt-gen/src/parser/qobject.rs]: Done, but we will need it at some point :-)

File: cxx-qt-gen/src/parser/qobject.rs (hunk 1 above)
Comment (ahayzen-kdab) [reply cxx-qt-gen/src/parser/qobject.rs]: Changed the wording for all of them.

File: cxx-qt-gen/src/parser/qobject.rs (hunk 3 above)
Comment (ahayzen-kdab) [reply cxx-qt-gen/src/parser/qobject.rs]: Done :-)

**Review by Be-ing — APPROVED** [review]

**Pull request merged**